### ISSUE TITLE
Initial CUDA 11.3 Conda Packages

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/cudatoolkit-feedstock#62]
We have published some new Conda packages for public consumption. These contain the redistributable libraries, compilers, profiling tools, etc. Also they are currently in the `nvidia` channel ( https://anaconda.org/nvidia ). To get started one can just run `conda install -c nvidia cuda=11.3`. This would include everything that `cudatoolkit` contains today.

We would like to collect some feedback from the community here to inform how we package these going forward. Once we are more sure these fill the needs here, we can follow up on integrating them into conda-forge.

**Comment by jakirkham** [comment]
Just to add the ARM packages are for [SBSA]( https://en.wikipedia.org/wiki/Server_Base_System_Architecture ). So will **not** work on Jetson for example

**Comment by jaimergp** [comment]
Nice, I'll have a look now. 

First thing I notice, the version strings are sometimes different across components:

```
$> CONDA_SUBDIR="linux-64" mamba create -n cuda cuda=11.3 -c nvidia
...

  cuda                           11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-command-line-tools        11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-compiler                  11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-cudart                   11.3.58  hc1aae59_0  nvidia/linux-64       1 MB
  cuda-cuobjdump                11.3.58  hc78e225_0  nvidia/linux-64     115 KB
  cuda-cupti                    11.3.58  h9a3dd33_0  nvidia/linux-64      19 MB
  cuda-cuxxfilt                 11.3.58  he670d9e_0  nvidia/linux-64      32 KB
  cuda-gdb                      11.3.58  h531059a_0  nvidia/linux-64      39 MB
  cuda-libraries                 11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-libraries-dev             11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-memcheck                 11.3.58  h8711ecb_0  nvidia/linux-64     156 KB
  cuda-nvcc                     11.3.58  h2467b9f_0  nvidia/linux-64      54 MB
  cuda-nvdisasm                 11.3.58  hd2ea46e_0  nvidia/linux-64      32 MB
  cuda-nvml-dev                 11.3.58  h70090ce_0  nvidia/linux-64      62 KB
  cuda-nvprof                   11.3.58  h860cd9e_0  nvidia/linux-64       4 MB
  cuda-nvprune                  11.3.58  hb917323_0  nvidia/linux-64      47 KB
  cuda-nvrtc                    11.3.58  he300756_0  nvidia/linux-64      30 MB
  cuda-nvtx                     11.3.58  h3fa534a_0  nvidia/linux-64      44 KB
  cuda-nvvp                     11.3.58  hd16380c_0  nvidia/linux-64     115 MB
  cuda-runtime                   11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-samples                  11.3.58  hc6eff01_0  nvidia/linux-64      65 MB
  cuda-sanitizer-api            11.3.58  h58da6c8_0  nvidia/linux-64      15 MB
  cuda-thrust                   11.3.58  h7b74f08_0  nvidia/linux-64       1 MB
  cuda-toolkit                   11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-tools                     11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  cuda-visual-tools              11.3.0  h3b286be_0  nvidia/linux-64       2 KB
  libcublas                11.4.2.10064  h8a72295_0  nvidia/linux-64     375 MB
  libcufft                    10.4.2.58  h58ccd86_0  nvidia/linux-64     358 MB
  libcurand                   10.2.4.58  h99380db_0  nvidia/linux-64      98 MB
  libcusolver                 11.1.1.58  hec68242_0  nvidia/linux-64     150 MB
  libcusparse                 11.5.0.58  hf5aa513_0  nvidia/linux-64     294 MB
  libnpp                      11.3.3.44  h8df316f_0  nvidia/linux-64     202 MB
  libnvjpeg                   11.4.1.58  h3d06750_0  nvidia/linux-64       4 MB
```

I see 11.3, 11.4, 10.4 (?). 


It'd be nice to see the underlying recipe(s). I guess they are included in the metadata, but it would be more convenient to have a Gist or something like that, if possible.

**Comment by jaimergp** [comment]
The dependency tree, if anybody is curious:

```
$> CONDA_SUBDIR=linux-64 mamba repoquery depends -q -c nvidia --tree cuda

Executing the query cuda

cuda[11.3.0]
  ├─ cuda-runtime[11.3.0]
  │  └─ cuda-libraries[11.3.0]
  │     ├─ cuda-cudart[11.3.58]
  │     ├─ cuda-nvrtc[11.3.58]
  │     ├─ libcublas[11.4.2.10064]
  │     ├─ libcufft[10.4.2.58]
  │     ├─ libcurand[10.2.4.58]
  │     ├─ libcusolver[11.1.1.58]
  │     ├─ libcusparse[11.5.0.58]
  │     ├─ libnpp[11.3.3.44]
  │     └─ libnvjpeg[11.4.1.58]
  └─ cuda-toolkit[11.3.0]
     ├─ cuda-compiler[11.3.0]
     │  ├─ cuda-cuobjdump[11.3.58]
     │  ├─ cuda-cuxxfilt[11.3.58]
     │  ├─ cuda-nvcc[11.3.58]
     │  └─ cuda-nvprune[11.3.58]
     ├─ cuda-libraries already visited
     ├─ cuda-libraries-dev[11.3.0]
     │  └─ cuda-thrust[11.3.58]
     ├─ cuda-nvml-dev[11.3.58]
     ├─ cuda-samples[11.3.58]
     └─ cuda-tools[11.3.0]
        ├─ cuda-command-line-tools[11.3.0]
        │  ├─ cuda-cupti[11.3.58]
        │  ├─ cuda-gdb[11.3.58]
        │  ├─ cuda-memcheck[11.3.58]
        │  ├─ cuda-nvdisasm[11.3.58]
        │  ├─ cuda-nvprof[11.3.58]
        │  ├─ cuda-nvtx[11.3.58]
        │  └─ cuda-sanitizer-api[11.3.58]
        └─ cuda-visual-tools[11.3.0]
           ├─ cuda-libraries-dev already visited
           ├─ cuda-nvml-dev already visited
           └─ cuda-nvvp[11.3.58]
```

**Comment by jakirkham** [comment]
Thanks Jaime! Would be great to have your feedback 😄

Also cc-ing @h-vetinari who may also be interested in these and have thoughts 🙂

**Comment by jakirkham** [comment]
> First thing I notice, the version strings are sometimes different across components

These are multiple separate recipes. So it makes sense that these differ. These would probably live as different feedstocks unless there is a reason to do something different

**Comment by jaimergp** [comment]
How many feedstocks would that be? This includes several packages and a handful of metapackages as well, which might be difficult to maintain and keep in sync. I don't know if it would be better to have them all in the same recipe, with multiple outputs, or if that would get extra complicated (I need to take a look at the recipes to see how bad the situation would get in a single meta.yaml).

The current split makes sense to me. I like that users can choose different types of granularity. It'd be nice to have the `cudatoolkit-major` and `cudatoolkit-major-minor` metapackages too, so maintainers can choose the level of ABI compat they want to keep, as per #48. (Although with the current subdivision, they could even choose they individual components they depend on so... maybe less of a reason?).

Also, is `nvcc` now redistributable? Or conda-forge wouldn't have access to that one?

**Comment by jakirkham** [comment]
AIUI combining them into a single recipe would be complicated.

Is that still needed with CUDA Enhanced Compatibility?

My understanding is all of this would be available to conda-forge. I don't know about general redistributability.

**Comment by jaimergp** [comment]
Having `nvcc` as a conda package would simplify parts of the setup, but I guess we still need the drivers in place.

I guess we'll have to do with several feedstocks, and have `cudatoolkit-feedstock` as the main hub for discussions.

**Comment by jaimergp** [comment]
I am trying out the packages to build OpenMM locally (maybe I'll put a PR to showcase how it's working once I get there).

One thing I noticed is that `cuda-nvcc` is not platform-suffixed, so `compiler('cuda')` won't work because conda will try to find `cuda-nvcc_linux-64` :/

**Comment by jaimergp** [comment]
Ok, so I have this working with OpenMM / 11.3 / Linux-64! Check https://github.com/conda-forge/openmm-feedstock/pull/60. It's  failing only because I am using `mambabuild` to speed up the debugging.

A few notes:

* OpenMM does not use CMake's lang features, only `FindCUDAToolkit`.
* For the autodetection to work, I needed to add both `cuda-nvcc` and `cuda-cudart` to `build`. No extra CMake / env variables were needed. It was all automatic!
* `host` needed `libcufft`, `cuda-cudart`, `cuda-nvrtc`, `cuda-nvprof`.
* The packages are not setting `run_exports` at all, so these need to be added manually to `run`.  `conda-build` will add helpful warnings otherwise.
* I used the 11.2 Docker image. I believe this is not interfering too much (just providing a recent enough driver I guess?), since the CUDA location found is `BUILD_PREFIX`, version is `11.3.58`, and the build does not work if I remove the `cuda-*` packages from `BUILD_PREFIX`.

So, in short, the feedback for now is super positive (at least for OpenMM), but we need `run_exports` definitions. I'd recommend that CUDA-heavier projects take this example and adapt it in their feedstocks in a WIP PR to confirm my findings.

**Comment by jaimergp** [comment]
Trying out Windows now. Initial feedback below:

* The contents tree uses the root `PREFIX` directly. Previous `cudatoolkit` packages used `PREFIX\Library` aka `%LIBRARY_PREFIX%`. This causes some issues with the expected paths.

**Comment by jaimergp** [comment]
One more thought, looks like CMake needs both `nvcc` and `cudart` to be present in the same PREFIX to detect a CUDA location accurately. In that case, we might just leave the packages named as they are (e.g. no need to have `cuda-nvcc_linux-64` and the like), but just have our `nvcc` wrapper point to these dependencies for CUDA 11.3+? For CUDA <=11.2, we would still use the wrapper?

**Comment by jakirkham** [comment]
It might make sense to do away with building against every minor version at the same time we move to these packages and use [CUDA Enhanced Compatibility]( https://docs.nvidia.com/deploy/cuda-compatibility/index.html ) instead. This will cutdown the amount of rebuilding we need to do while still supporting CUDA 11.0+

**Comment by jakirkham** [comment]
cc @seibert

**Comment by jaimergp** [comment]
@jakirkham Is there anything else you want me test/try here?

**Comment by jaimergp** [comment]
Oh, by the way, I forgot to mention that thanks to these packages, the Windows setup script for CUDA is not needed! Note how it fails because 11.3 is not [available](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=358224&view=logs&jobId=171a126d-c574-5c8c-1269-ff3b989e923d&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=11e12034-2a2b-52f8-fa15-1d2b32e6923f) yet there and still the build [succeeds](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=358224&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=5354) for OpenMM.

**Comment by leofang** [comment]
@jakirkham It seems `cuda_profiler_api.h` is missing? At least it's not in `cuda-cudart` or `cuda-runtime`.

**Comment by leofang** [comment]
Ah, it's `cuda-nvprof`.

**Comment by leofang** [comment]
Hi @jakirkham, looks like the entire folder `/usr/local/cuda/include/crt` is not included in `cuda-cudart` as it should be?

**Comment by leofang** [comment]
It seems CuPy can be compiled with the new package layout on linux64 and windows. The steps I took is similar to @jaimergp's tests for OpenMM but with some differences. Here it is:

For CBC:
1. Add `nvidia` to `channel_sources`
2. For linux64, use the non-CUDA cos7 docker image (to avoid accidentally using non-conda CTK)
3. Change `nvcc` to `cuda_nvcc` in `cuda_compiler`
4. Add CUDA 11.3 (to ensure CTK comes from `nvidia` and not from `conda-forge`)

For meta.yaml:
1. For linux64, append `-L${PREFIX}/lib/stubs` to `CFLAGS` (to let the linker find the `libcuda.so` stub, as I do not use any docker image with CUDA pre-installed)
2. For linux64, append `-ccbin $CXX` to `nvcc` so that it knows we're using conda-forge's compiler (this is a job done in nvcc-feedstock and should also be done in `cuda-nvcc` [1]).
3. Set `CUDA_PATH` to `$PREFIX` (again, a job done in nvcc-feedstock)
4. Temporarily remove all optional "satellite" CUDA dependencies like cuDNN, NCCL, etc (they need to be rebuilt against the new CUDA package to fix their dependencies)
5. Remove `{{ compiler("cuda") }}` in the build requirement [2]
6. Add `cuda-nvcc` and `cuda-cudart` to the host requirement [2]
7. Add needed CUDA components to the host section (in CuPy's case this includes `cuda-cudart`, `cuda-nvrtc`, `cuda-nvprof` and all math libraries) [3]

[1]: On Windows, nvcc for some reason picks up the host compiler correctly, but I think it's a bit suspicious considering the CI setup might do something when installing CUDA in the CI runner...

[2]: I had a setup like this
```yaml
build:
  - ...
  - cuda-nvcc

host:
  - cuda-nvcc
  - cuda-cudart
  - ...
```
but apparently the nvcc from build's `cuda-nvcc` alone cannot find `cuda_runtime.h` from host's `cuda-cudart` so they must come in pair. I bet they can both be moved to the build section. Another reason that they must come in pair: without `cuda-cudart` the CUDA runtime headers in `include/crt` cannot be found. I believe this makes the naive substitution `s/{{ compiler("nvcc") }}/cuda-nvcc` difficult. 

[3]: Currently there's no version pinning of each CUDA components to its parent CTK version. So, for example, when I request `libcufft` the one installed is actually from CTK 11.4.1, not from 11.3.x. I can, of course, find out the exact version that comes with CTK 11.3.0 and use this info, but it's too tedious for downstream feedstock maintainers and I consider it a bug to be fixed.

Note: I haven't checked if all CUDA components have the correct `run_exports` -- I suspect they don't, given @jaimergp's discussions above.

cc: @kmaehashi @jakirkham

**Comment by leofang** [comment]
> I haven't checked if all CUDA components have the correct `run_exports` -- I suspect they don't

Confirmed `run_exports` is not correctly set up, so during the test phase `libcudart.so` cannot be found:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=366350&view=logs&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642&l=6833
Looks like among all the components only `cuda-nvcc` has `run_exports`?

**Comment by jaimergp** [comment]
Yes, let me summarize my understanding here, @jakirkham! @leofang, could you please review and let me know if I missed something?

First, the positive aspects: with these packages there's a high chance we will _not_ need a Docker image to provide a system CUDA! This also applies to Windows, where no ad-hoc installation will be needed. This will save bandwidth, disk space and time! 💪 

__General__

* As a compiler package, `cuda-nvcc` needs to be suffixed with the platform; e.g. `cuda-nvcc_linux-64`.
* The `build` prefix needs both `cuda-nvcc` _and_ `cuda-cudart`. 
  * I find this a bit surprising, since I would expect that redefining `compiler('cuda')` to point to `cuda-nvcc` should be enough. However, without `cuda-cudart`, `cuda_runtime.h` and `include/crt` headers cannot be found (even if provided in `host`). 
  * If these need to be kept separate, a metapackage might be needed to wrap them under the `compiler()` call. We do have `cuda-compiler` but it includes more stuff than needed, without including `cuda-cudart`.
* Set some flags and environment variables to make it `conda-build` friendly. An `activate.d` script similar to the one we use in our current wrapper needs to be provided with `cuda-nvcc` (or whatever we end up pointing to `compiler('cuda')`). It should:
  * Set `-ccbin` to our compilers (is there an env var for this?)
  * Set `CUDA_PATH=$PREFIX`
  * Add `-L$PREFIX/lib/stubs` to `CFLAGS`
* `run_exports` definitions have to be revised for all packages. Most are _not_ exporting their runtime dependencies.
* Versions are not pinned to their parent `cudatoolkit`. As @leofang said:
> Currently there's no version pinning of each CUDA components to its parent CTK version. So, for example, when I request libcufft the one installed is actually from CTK 11.4.1, not from 11.3.x. I can, of course, find out the exact version that comes with CTK 11.3.0 and use this info, but it's too tedious for downstream feedstock maintainers and I consider it a bug to be fixed.

__Windows-only__

* The packages are extracted to `$PREFIX`, but, usually, unix-style contents are extracted to `$PREFIX/Library` on Windows.

**Comment by leofang** [comment]
Thanks @jaimergp for great summary! I don't think you missed anything 👍

**Comment by jakirkham** [comment]
@soumith, would you be able to look at these as well and share your feedback? There are CUDA 11.5 packages too

**Comment by jakirkham** [comment]
Also if others here can take a look at the 11.5 packages and see if those have addressed your comments or if there are others things we should look at, that would be very helpful 🙂

**Comment by jaimergp** [comment]
Nice! I'll take a look when I have some time and run the OpenMM builds with the new stack again.

For now, the dependency tree:

```
$> CONDA_SUBDIR=linux-64 mamba repoquery depends -q -c nvidia --tree cuda=11.5

Executing the query cuda=11.5



cuda[11.5.0]
  ├─ cuda-runtime[11.5.0]
  │  └─ cuda-libraries[11.5.0]
  │     ├─ cuda-cudart[11.5.50]
  │     ├─ cuda-nvrtc[11.5.50]
  │     ├─ libcublas[11.7.3.1]
  │     ├─ libcufft[10.6.0.54]
  │     ├─ libcurand[10.2.6.48]
  │     ├─ libcusolver[11.2.1.48]
  │     ├─ libcusparse[11.7.0.31]
  │     ├─ libnpp[11.5.1.53]
  │     └─ libnvjpeg[11.5.3.48]
  └─ cuda-toolkit[11.5.0]
     ├─ cuda-compiler[11.5.0]
     │  ├─ cuda-cuobjdump[11.5.50]
     │  ├─ cuda-cuxxfilt[11.5.50]
     │  ├─ cuda-nvcc[11.5.50]
     │  └─ cuda-nvprune[11.5.50]
     ├─ cuda-libraries already visited
     ├─ cuda-libraries-dev[11.5.0]
     │  ├─ cuda-cudart-dev[11.5.50]
     │  │  ├─ cuda-cccl[11.5.62]
     │  │  └─ cuda-cudart already visited
     │  ├─ cuda-driver-dev[11.5.50]
     │  ├─ cuda-nvrtc-dev[11.5.50]
     │  │  └─ cuda-nvrtc already visited
     │  ├─ libcublas-dev[11.7.3.1]
     │  │  └─ libcublas already visited
     │  ├─ libcufft-dev[10.6.0.54]
     │  │  └─ libcufft already visited
     │  ├─ libcurand-dev[10.2.6.48]
     │  │  └─ libcurand already visited
     │  ├─ libcusolver-dev[11.2.1.48]
     │  │  └─ libcusolver already visited
     │  ├─ libcusparse-dev[11.7.0.31]
     │  │  └─ libcusparse already visited
     │  ├─ libnpp-dev[11.5.1.53]
     │  │  └─ libnpp already visited
     │  └─ libnvjpeg-dev[11.5.3.48]
     │     └─ libnvjpeg already visited
     ├─ cuda-nvml-dev[11.5.50]
     ├─ cuda-samples[11.5.56]
     │  ├─ cuda-libraries-dev[11.4.1]
     │  │  └─ cuda-thrust[11.4.43]
     │  └─ cuda-nvcc already visited
     └─ cuda-tools[11.5.0]
        ├─ cuda-command-line-tools[11.5.0]
        │  ├─ cuda-cupti[11.5.57]
        │  ├─ cuda-gdb[11.5.50]
        │  ├─ cuda-memcheck[11.5.50]
        │  ├─ cuda-nvdisasm[11.5.50]
        │  ├─ cuda-nvprof[11.5.50]
        │  ├─ cuda-nvtx[11.5.50]
        │  └─ cuda-sanitizer-api[11.5.50]
        └─ cuda-visual-tools[11.5.0]
           ├─ cuda-libraries-dev already visited
           ├─ cuda-nvml-dev already visited
           └─ cuda-nvvp[11.5.50]
              ├─ cuda-nvdisasm[11.3.58]
              └─ cuda-nvprof already visited
```

I guess it's ok that older versions are kept in newer versions, but how come there are some packages coming from 11.7?

**Comment by jaimergp** [comment]
Opened https://github.com/conda-forge/openmm-feedstock/pull/64 to investigate. I'll keep you posted!

**Comment by jakirkham** [comment]
Thanks Jaime! 😄

**Comment by jakirkham** [comment]
cc @hmaarrfk (for awareness)

**Comment by leofang** [comment]
> I guess it's ok that older versions are kept in newer versions, but how come there are some packages coming from 11.7?

@jaimergp no that's each library's own version, independent of CTK's version.

**Comment by kkraus14** [comment]
Hey all, is there any updates or plans on upstreaming the new packages that are being produced in the `nvidia` channel to conda forge? They're a huge quality of life improvement versus the current state in conda-forge!

**Comment by jakirkham** [comment]
@jaimergp would you have time to look at the 11.6 packages on the `nvidia` channel and let us know if your concerns have been addressed? 🙂

**Comment by jaimergp** [comment]
I'll try next week!

**Comment by jakirkham** [comment]
@jaimergp have you had a chance to look? If not, no worries 🙂

**Comment by jaimergp** [comment]
I opened a PR on the openmm-feedstock as usual: https://github.com/conda-forge/openmm-feedstock/pull/69

It's just the same we had for 11.5, but using 11.6. Let's see if we can clean it up!

**Comment by hansenms** [comment]
@jakirkham, I have just randomly bounced into this thread. I have been using the cuda packages from the nvidia channel for a little bit but recently (yesterday) the 11.6.0 subchannel (label) packages broke. I wrote this post on the nvidia developer forum: https://forums.developer.nvidia.com/t/conda-packages-from-the-nvidia-label-cuda-11-6-0-failing-to-install/204281 but I am not actually sure anybody is picking it up. Where is the right place to provide feedback on these packages?

**Comment by jakirkham** [comment]
@hansenms Keeping that discussion in that developer forum seems like the right choice atm

This thread is about integrating those packages into conda-forge as well as periodic testing to confirm issues raised previously have been addressed

**Comment by hansenms** [comment]
@jakirkham I will see if somebody has thoughts or a response on the developer forum. If you have any contacts you can direct there, it would be much appreciated. I would at least like to understand it a bit better. Thanks.

**Comment by jakirkham** [comment]
I've already asked someone internally to take a look 🙂

**Comment by hansenms** [comment]
@jakirkham, I hate to ping you on this again, but nobody has responded on that thread on the nvidia developer forum. Is there a better way to reach the team that make the conda packages in the nvidia channel?

**Comment by hansenms** [comment]
@jakirkham not sure if there is something that you could do, but I would really like to know where to post an issue/bug/problem with the official nvidia conda packages if it is not on the developer forum (where there is no response) and not here. Any good suggestions for where to raise the issue would be much appreciated.

**Comment by hansenms** [comment]
@jakirkham the problem mentioned above just happened overnight again for the cuda 11.6.1 packages. It is just massively disruptive that these packages keep getting broken overnight and nobody is responding to the issue raised on the nvidia developer forum. Where can one go to raise an issue about these packages?

**Comment by leofang** [comment]
Looks like the old packages are removed whenever a new package is uploaded.

**Comment by hansenms** [comment]
@leofang yes, but that is massively disruptive. And nvidia's own instructions are broken, so for instance:

```
conda install -c nvidia/label/cuda-11.6.1 cuda-toolkit
```

Will not work because they removed the samples package and other packages in that label.

**Comment by ngam** [comment]
@jakirkham let's get this done 🚀 🌕 🙌 💎 

boom boom, buzz buzz

**Comment by ngam** [comment]
Is there anything non-nvidia people can do to help get this done? Are there any blockers or can we get an update on the status of this? Thanks all for the great work!

**Comment by jakirkham** [comment]
As we are now doing this work and it is being tracked in issue ( https://github.com/conda-forge/staged-recipes/issues/21382 ), closing this issue in favor of discussion there and reviews of PRs linked from that staged-recipes issue. Thanks everyone! 🙏
